crates/turborepo-cache/src/signature_authentication.rs

```diff
@@ -0,0 +1,267 @@
+use std::env;
+
+use base64::{prelude::BASE64_STANDARD, Engine};
+use os_str_bytes::OsStringBytes;
+use ring::{
+    hmac,
+    hmac::{Tag, HMAC_SHA256},
+};
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum SignatureError {
+    #[error(
+        "signature secret key not found. You must specify a secret key in the \
+         TURBO_REMOTE_CACHE_SIGNATURE_KEY environment variable"
+    )]
+    NoSignatureSecretKey,
+    #[error("serialization error: {0}")]
+    SerializationError(#[from] serde_json::Error),
+    #[error("base64 encoding error: {0}")]
+    Base64EncodingError(#[from] base64::DecodeError),
+}
+
+#[derive(Debug)]
+pub struct ArtifactSignatureAuthenticator {
+    team_id: Vec<u8>,
+    // An override for testing purposes (to avoid env var race conditions)
+    secret_key_override: Option<Vec<u8>>,
+}
+
+impl ArtifactSignatureAuthenticator {
+    pub fn new(team_id: Vec<u8>, secret_key_override: Option<Vec<u8>>) -> Self {
+        Self {
+            team_id,
+            secret_key_override,
+        }
+    }
+
+    fn secret_key(&self) -> Result<Vec<u8>, SignatureError> {
```

Comment:
I believe that HMAC_SHA256 doesn't have key length requirements but a comment that verifies that would be handy.